osu.Game.Rulesets.Osu/Skinning/Legacy/LegacySpinner.cs

```diff
@@ -127,5 +129,33 @@ protected override void Dispose(bool isDisposing)
             if (DrawableSpinner != null)
                 DrawableSpinner.ApplyCustomUpdateState -= UpdateStateTransforms;
         }
+
+        /// <summary>
+        /// A <see cref="Container"/> simulating osu!stable's absolute screen-space,
+        /// for perfect placements of legacy spinner components with legacy coordinates.
+        /// </summary>
+        protected class LegacyCoordinatesContainer : Container
+        {
+            /// <summary>
+            /// An offset that simulates stable's spinner top offset,
+            /// for positioning some legacy spinner components perfectly as in stable.
+            /// (e.g. 'spin' sprite, 'clear' sprite, metre in old-style spinners)
+            /// </summary>
+            public static readonly float SPINNER_TOP_OFFSET = (float)Math.Ceiling(45f * SPRITE_SCALE);
+
+            public LegacyCoordinatesContainer()
+            {
+                // legacy spinners relied heavily on absolute screen-space coordinate values.
+                // wrap everything in a container simulating absolute coords to preserve alignment
+                // as there are skins that depend on it.
+                Anchor = Anchor.Centre;
+                Origin = Anchor.Centre;
+                Size = new Vector2(640, 480);
+
+                // since legacy coordinates were on screen-space, they were accounting for the playfield shift offset.
+                // therefore cancel it from here.
+                Position = new Vector2(0, -8f);
```

Comment:
comment may not make much sense, but not sure how to better word it..

Reply:
```suggestion
                // counteracts the playfield shift from OsuPlayfieldAdjustmentContainer.
                Position = new Vector2(0, -8f);
```

also note that this alignment is *off* in edit mode. so i'm not sure whether things will continue to look fine there.

Reply:
seems like breaking the legacy coordinates is the way here. instead of `115` and `335`, needs to be forced into `115 - 8` and `335 - 8`.

Reply:
I don't understand what you're saying.

Reply:
actually no, nevermind me. this is correct the way it is, alignment isn't off in edit mode. what this code is doing is counteract the playfield shift in the legacy coordinates themselves (`115` for example has playfield shift applied to it, osu!stable subtracts from `spinner_top_offset` to counteract it)

```csharp
// counteracts the playfield shift applied to the legacy coordinates.
Position = new Vector2(0, -8f);
```

not sure whether this comment explains it, what do you think?

Reply:
> alignment isn't off in edit mode

I beg to differ? Compare [`DrawableOsuRuleset`](https://github.com/ppy/osu/blob/9b5d11f2a521e41c4fdf282e6e58905aa52f8669/osu.Game.Rulesets.Osu/UI/DrawableOsuRuleset.cs#L42) and [`DrawableOsuEditRuleset`](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Osu/Edit/DrawableOsuEditRuleset.cs#L26). That's the same check as the `CorrectionOffsetActive` thing in stable.

Reply:
this has nothing to do with whether the playfield shift is applied to the playfield adjustment container.

this code counteracts/cancels/subtracts/whatever osu!stable's playfield shift from the legacy coordinates, `spinner_top_offset + 115` for example has playfield shift applied to it, osu!stable subtracts from `spinner_top_offset` to remove that shift when in edit mode (`CorrectionOffsetActive`).

I'm not sure how to explain this any further than that to be honest, I'm out of words.

Reply:
ok, let's put it in a different way. stable has a conditional hinging on `CorrectionOffsetActive`. your code has no such thing. why?

Reply:
because it is applied at an `OsuPlayfieldAdjustmentContainer` level, which `LegacyCoordinatesContainer` respects, by:
https://github.com/ppy/osu/blob/0ad3073c1aa3863edbfb4c4f485ad970e507127e/osu.Game.Rulesets.Osu/Skinning/Legacy/LegacySpinner.cs#L146-L152

that's why this container completely counteracts the playfield shift from legacy coordinates, to make them respect the adjustment container's applied playfield shift.

Reply:
well you see, *that* is an actual understandable explanation, and is the metric for what a good comment here should read. the key point is that the adjustment is handled at a higher level, so it's safe to assume that the shift is always on in this component.

fwiw I came at this conclusion independently before reading your latest comment. but it took way too long to surmise that from your previous attempts to explain.

Reply:
so what do you think about the inline comment, currently it's not quite correct. maybe this:
```csharp
// playfield shift is applied at a higher level, counteract it from legacy coordinates here.
Position = new Vector2(0, -8f);
```

Reply:
```csharp
// stable applies this adjustment conditionally, locally in the spinner.
// in lazer this is handled at a higher level in OsuPlayfieldAdjustmentContainer,
// therefore it's safe to apply it unconditionally in this component.
```